File: yangxj96-serve/yangxj96-serve-auth/src/main/java/com/yangxj96/saas/server/auth/service/impl/RoleServiceImpl.java
package com.yangxj96.saas.server.auth.service.impl;

import cn.hutool.core.lang.tree.Tree;
import cn.hutool.core.lang.tree.TreeNodeConfig;
import cn.hutool.core.lang.tree.TreeUtil;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.yangxj96.saas.bean.user.Authority;
import com.yangxj96.saas.bean.user.Role;
import com.yangxj96.saas.common.base.BaseServiceImpl;
import com.yangxj96.saas.common.respond.R;
import com.yangxj96.saas.common.respond.RStatus;
import com.yangxj96.saas.server.auth.mapper.RoleMapper;
import com.yangxj96.saas.server.auth.pojo.vo.RoleRelevance;
import com.yangxj96.saas.server.auth.service.RoleService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;

/**
 * 角色service的实现
 */
@Service
public class RoleServiceImpl extends BaseServiceImpl<RoleMapper, Role> implements RoleService {

    protected RoleServiceImpl(RoleMapper bindMapper) {
        super(bindMapper);
    }

    @Transactional(rollbackFor = Exception.class)
    public Role create(Role datum) {
        if (!datum.getCode().startsWith("ROLE_")) {
            R.failure(RStatus.FAILURE_FORMAT);
            return datum;
        }
        var wrapper = new QueryWrapper<Role>()
                .eq("code", datum.getCode())
                .last("LIMIT 1");
        var db = this.getOne(wrapper);
        if (this.getOne(wrapper) != null) {
            R.failure(RStatus.FAILURE_REPEAT);
            return db;
        }
        return super.create(datum);
    }

    @Transactional(rollbackFor = Exception.class)
    public void relevance(RoleRelevance params) {

    }

    public List<Tree<String>> tree() {
        var roles = this.list();
        if (roles.isEmpty()) {
            return Collections.emptyList();
        }
        var config = new TreeNodeConfig();
        config.setIdKey("id");
        config.setParentIdKey("pid");
        config.setNameKey("name");
        config.setChildrenKey("children");
        return TreeUtil.build(roles, "0", config, (node, tree) -> {
            tree.setId(node.getId().toString());
            tree.setParentId(node.getPid().toString());
            tree.setName(node.getName());
            tree.putExtra("code", node.getCode());
            tree.putExtra("description", node.getDescription());
        });
    }

    public List<Authority> ownerAuthority(Long id) {
        return bindMapper.ownerAuthority(id);
    }

    public List<Authority> getAuthorityByRoleCode(String roleCode) {
        var wrapper = new LambdaQueryWrapper<Role>()
                .eq(Role::getCode, roleCode)
                .last("LIMIT 1");
        var role = this.getOne(wrapper);
        if (role == null) {
            return Collections.emptyList();
        }
        return bindMapper.getAuthorityByRoleCode(role.getId());
    }
}

File: yangxj96-serve/yangxj96-serve-auth/src/main/kotlin/com/yangxj96/saas/server/auth/service/impl/RoleServiceImpl.kt
package com.yangxj96.saas.server.auth.service.impl

import cn.hutool.core.lang.tree.Tree
import cn.hutool.core.lang.tree.TreeNodeConfig
import cn.hutool.core.lang.tree.TreeUtil
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.baomidou.mybatisplus.core.metadata.IPage
import com.baomidou.mybatisplus.extension.kotlin.KtQueryWrapper
import com.baomidou.mybatisplus.extension.plugins.pagination.Page
import com.yangxj96.saas.bean.user.Authority
import com.yangxj96.saas.bean.user.Role
import com.yangxj96.saas.common.base.BaseServiceImpl
import com.yangxj96.saas.common.respond.R
import com.yangxj96.saas.common.respond.RStatus
import com.yangxj96.saas.server.auth.mapper.RoleMapper
import com.yangxj96.saas.server.auth.pojo.vo.RoleRelevance
import com.yangxj96.saas.server.auth.service.RoleService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 角色service的实现
 */
@Service
class RoleServiceImpl protected constructor(bindMapper: RoleMapper) :
    BaseServiceImpl<RoleMapper, Role>(bindMapper), RoleService {

    @Transactional(rollbackFor = [Exception::class])
    override fun create(datum: Role): Role {
        if (!datum.code?.startsWith("ROLE_")!!) {
            R.failure(RStatus.FAILURE_FORMAT)
            return datum
        }

        val wrapper = QueryWrapper<Role>()
            .eq("code", datum.code)
            .last("LIMIT 1")

        val db = this.getOne(wrapper)
        if (this.getOne(wrapper) != null) {
            R.failure(RStatus.FAILURE_REPEAT)
            return db!!
        }
        return super.create(datum)
    }

    @Transactional(rollbackFor = [Exception::class])
    override fun relevance(params: RoleRelevance) {

    }

    override fun tree(): List<Tree<String>> {
        val roles = this.list()
        if (roles.isNullOrEmpty()) {
            return mutableListOf()
        }
        val config = TreeNodeConfig()
        config.setIdKey("id")
        config.setParentIdKey("pid")
        config.setNameKey("name")
        config.setChildrenKey("children")
        return TreeUtil.build(roles, "0", config) { node: Role, tree: Tree<String> ->
            tree.setId(
                node.id.toString()
            )
            tree.setParentId(node.pid.toString())
            tree.setName(node.name)
            tree.putExtra("code", node.code)
            tree.putExtra("description", node.description)
        }
    }

    override fun ownerAuthority(id: Long): MutableList<Authority> {
        return bindMapper.ownerAuthority(id)
    }

    override fun getAuthorityByRoleCode(roleCode: String): List<Authority> {
        val role = this.getOne(
            KtQueryWrapper(Role::class.java)
                .eq(Role::code, roleCode)
                .last("LIMIT 1")
        ) ?: return emptyList()
        return bindMapper.getAuthorityByRoleCode(role.id)
    }
}
